actionpack/lib/action_controller/metal/renderers.rb

```diff
@@ -103,17 +142,106 @@ def self.remove(key)
       remove_method(method_name) if method_defined?(method_name)
     end
 
+    # A Set containing serializer names that correspond to available serializer procs.
+    # Default values are <tt>:json</tt>, <tt>:js</tt>, <tt>:xml</tt>.
+    SERIALIZERS = Set.new
+
+    def self._serialize_with_serializer_method_name(key)
+      "_serialize_with_serializer_#{key}"
+    end
+
+    # Serializers define a method called within a renderer specific to
+    # transforming the object into a mime-compatible type.
+    # See <tt>Renderers.add</tt>
+    #
+    # The separation of serialization from rendering allows
+    # composing the Renderer behavior of two methods, e.g.
+    # +_render_with_renderer_json+ and +_serialize_with_serializer_json+,
+    # rather than requiring one to define a method +_render_with_renderer_json+
+    # in a subclass and optionally call super on it.
+    #
+    # A principal benefit of this approach is that it promotes serialization of an object
+    # to a clearly-defined public interface, rather than requiring one to understand that
+    # calling, e.g. +render json: object+ calls +_render_to_body_with_renderer(options)+
+    # which calls +_render_with_renderer_#{key}+ where key is +json+, which is the method
+    # defined by calling +ActionController::Renderers.add :json+.
+    #
+    # Example usage:
+    #
+    # Prior to the introduction of SERIALIZERS, the ActiveModel::Serializers gem,
+    # for example, has relied upon defining  +_render_option_json+ ( pre-4.2 )
+    # and +_render_with_renderer_json+ in the controller, and calling +super+ on
+    # the serialized object. Now, ActiveModel::Serializers need only call
+    # +ActionController.remove_serializer :json+ and define a new serializer with
+    # +ActionController.add_serializer json do |json, options| end+. There's
+    # no longer a need to add controller methods to define custom serializers.
+    #
+    # Pretty-printing JSON can be implemented by replacing the JSON serializer:
+    #
+    #   ActionController::Renderers.remove_serializer :json
+    #   ActionController::Renderers.add_serializer :json do |json, options|
+    #     return json if json.is_a?(String)
+    #
+    #     json = json.as_json(options) if json.respond_to?(:as_json)
+    #     json = JSON.pretty_generate(json, options)
+    #   end
+    #
+    # The JSON serializer could be extended by overriding it:
+    #
+    #   def _serialize_with_serializer_json(json, options)
+    #     json = CustomSerializer.new(json, options)
+    #     super
+    #   end
+    #
```

Comment:
It's interesting that you can't really look up a renderer/serializer right now, that the dance Rails is doing is mapping e.g. `:json` to `_render_with_renderer_json(json, options)`, e.g. that to [render to json, it iterates, builds a method name, and sends to it](https://github.com/rails/rails/blob/8c2c1bccccb3ea1f22503a184429d94193a1d9aa/actionpack/lib/action_controller/metal/renderers.rb#L44-L45)

```ruby
    def render_to_body(options)
      _render_to_body_with_renderer(options) || super
    end

    def _render_to_body_with_renderer(options)
      _renderers.each do |name|
        if options.key?(name)
          _process_options(options)
          method_name = Renderers._render_with_renderer_method_name(name)
          return send(method_name, options.delete(name), options)
        end
      end
      nil
    end
```

That there's no `Renderers.for(:json).render(json, options)` which might use an object like `JsonRenderer` or even the proc/block defined as the body of `_render_with_renderer_json`.

And having [ActionController::Rendering.renderer](https://github.com/rails/rails/blob/fecf9d729a81c6c08fef92936e659cdfcf340be5/actionpack/lib/action_controller/metal/rendering.rb#L16-L18) call [ActionController::Renderer.for(controller)](https://github.com/rails/rails/blob/3e36db4406beea32772b1db1e9a16cc1e8aea14c/actionpack/lib/action_controller/renderer.rb#L42-L54) which creates and anonymous subclass of `ActionController::Renderer` it appears is only used for templates?